Step3 자동차 경주

pikachom opened the pull request:
안녕하세요 리뷰어님
자동차 경주에 대한 리뷰 및 PR요청 드립니다.
코드 이외에 드릴 질문으로,
private function에 대한 단위테스트는 하지 않는 것이 좋다고 하여 public function에 대해서만 테스트를 구현하였는데
이와 관련하여 의견 주시면 큰 도움이 될 것 같습니다. 
감사합니다

kyle-um committed:
study: add kotlin basic

kyle-um committed:
study: add kotlin basic

kyle-um committed:
[feature] StringCalculator 기능 구현

pikachom committed:
[test] StringCalculatorTest 구현

- 테스트 동작을 위한 run() 함수의 return type 추가
- StringCalculatorTest 구현(with KoTest)

pikachom committed:
Merge branch 'pikachom' into Step2-StringCalculator

pikachom committed:
doc: step3의 요구사항 정리

pikachom committed:
feat: 자동차 경주 구현

pikachom committed:
refactor: 함수 책임 분리 및 가독성을 위한 변경

pikachom committed:
test: 단위테스트 구현

pikachom committed:
lint: ktLintFormat 수행

pikachom committed:
Merge branch 'pikachom' into step3

pikachom committed:
refact: step2 리뷰 코멘트 반영(1)

https://github.com/next-step/kotlin-racingcar/pull/554#discussion_r873520006

- 매직넘버 제거, 문맥 전달을 위한 함수추출

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,35 @@
+    companion object {
+        val gameEngine = GameEngine
+        val carList = mutableListOf<Car>()
+    }
```
클래스 내부의 요소들은 다음과 같은 순서로 정렬되어야 합니다.

1. Property declarations and initializer blocks
2. Secondary constructors
3. Method declarations
4. Companion object

https://kotlinlang.org/docs/coding-conventions.html#class-layout

wisemuji replied:
추가로, 해당 두 필드가 companion object가 되어야 할 필요가 있을지 생각해보시면 좋을 것 같아요! 상태를 가지는 변수는 companion object에 적합하지 않을 수 있습니다.

pikachom replied:
두가지는 일종의 게임 서버를 생각하며 글로벌, 유니크한  인스턴스가 있어야 한다고 생각해서 companion object로 두었었는데요,
오늘 수업을 듣고 나니 굳이 요구조건에서 필요하지 않은 컨텍스트를 반영할 필요가 없는 것 같아 변경하도록 하겠습니다.

pikachom committed:
refact: step2 리뷰 코멘트 반영(2)

https://github.com/next-step/kotlin-racingcar/pull/554#discussion_r873521458

- operator 를 enum class 로 변경하고, 검증 부분 위임

wisemuji commented on src/main/kotlin/racing/Application.kt:
```diff
@@ -0,0 +1,28 @@
+package racing
+
+fun main() {
```
Application과 Game에 있는 View 관련 로직을 요구사항에 맞게 InputView와 ResultView로 분리해보시면 어떨까요?

![image](https://user-images.githubusercontent.com/32327475/168575248-b5ad695b-c90e-4c28-9113-06d7e8eda5ba.jpeg)

pikachom replied:
반영하였습니다 😄

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,35 @@
+
+class Game {
+    companion object {
+        val gameEngine = GameEngine
```
오브젝트 타입을 변수가 담고 있어야 할 필요가 있을까요?

pikachom replied:
첫번째 코멘트에 대한 답변과 마찬가지 이유로 변경하였습니다.

pikachom committed:
refact: step2 리뷰 코멘트 반영(3)

https://github.com/next-step/kotlin-racingcar/pull/554#discussion_r873525748

- run 역할 분리

wisemuji commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,10 @@
+
+data class Car(
+    val number: Int,
+    var currentLocation: Int = 0,
```
currentLocation 변수가 가변이지만 외부로 노출되어 있어서 수정이 가능해요. 가변 변수는 내부에서만 조작할 수 있도록 해보면 어떨까요?

pikachom replied:
꼭 필요하지 않았던 data class를 일반 class로 변경하고, 오늘 강의에서 배운 private set 키워드를 통해 수정하였습니다.

wisemuji commented on src/main/kotlin/racing/GameEngine.kt:
```diff
@@ -0,0 +1,18 @@
+    }
+    private fun playTurn(car: Car) {
+        val diceResult = getRandomNumberBetween(randomRange)
+        if (carMoveDecision(diceResult)) car.move()
```
이 로직은 어떻게 테스트해볼 수 있을까요?

pikachom replied:
private을 제거하고, 로직을 변경하여 테스트를 구현했습니다.

pikachom replied:
private을 제거하고, 로직을 변경하여 테스트를 구현했습니다.

wisemuji commented on src/test/kotlin/racing/CarTest.kt:
```diff
@@ -0,0 +1,32 @@
+            carInstance.move()
+            carInstance.currentLocation shouldBe 2
+            carInstance.move()
+            carInstance.currentLocation shouldBe 3
```
매번 이렇게 반복적으로 move 호출 후 currentLocation값을 비교해야 할까요?

pikachom replied:
반복문을 사용하여 move가 반복해서 나열된 부분을 변경하였습니다.

wisemuji commented on src/test/kotlin/racing/GameTest.kt:
```diff
@@ -0,0 +1,36 @@
+    val carNumberInput = 3
+    val turnNumberInput = 4
+    val game = Game()
+    context("run() 단위테스트") {
```
어떤 로직을 검증하려고 하는지 알기 어렵습니다.
1. 먼저 이 테스트를 통해 내가 검증하고자 하는 것이 무엇인지 생각해보고
2. 그걸 테스트명을 통해 드러내보면 어떨까요?

pikachom replied:
테스트명을 이해하기 쉽도록 변경하였습니다.

wisemuji reviewed (changes requested):
자동차 경주도 다양한 경우의 테스트를 꼼꼼히 작성해주셨군요! 💪
이번 단계에서는 어떻게 해야 랜덤한 로직을 테스트할 수 있을까에 대한 고민을 해보시면 좋을 것 같습니다 🙂

같이 고민해보시면 좋을 만한 코멘트를 남겼으니 궁금한 점이 있다면 편하게 DM 주세요!

wisemuji commented:
> 코드 이외에 드릴 질문으로,
private function에 대한 단위테스트는 하지 않는 것이 좋다고 하여 public function에 대해서만 테스트를 구현하였는데
이와 관련하여 의견 주시면 큰 도움이 될 것 같습니다.

@pikachom

먼저 private function에 대한 단위테스트가 필요한지 판단해보는것이 좋아요!

* 만약 해당 함수가 가독성의 이유로 분리되었다면: public function에 대한 테스트만으로도 검증 가능
* 가독성 이상의 역할을 하는 경우: 클래스 분리가 권장됨

만약 가독성 이상의 역할을 하는 경우 해당 역할/책임을 갖는 다른 객체를 만들 타이밍이 아닐지 고민 해 볼 수 있습니다.

pikachom committed:
refactor: calculate에서 가변변수(operator) 제거

https://github.com/next-step/kotlin-racingcar/pull/554#discussion_r873603369

pikachom committed:
refactor: 불필요한 companion object 제거

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873577469

pikachom committed:
refactor: UI로직 분리 구현

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873576258

pikachom committed:
refact: InputVie, ResultView 분리 대응 테스트 구현

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873576258

pikachom committed:
refact: Car를 data class에서 class로 변경

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873678384

pikachom committed:
refactor: playTurn 단위테스트 추가

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873679265

pikachom committed:
refactor: move 테스트 반복호출을 반복문으로 변경

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873680854

pikachom committed:
refactor: run 단위테스트 이름 변경

https://github.com/next-step/kotlin-racingcar/pull/568#discussion_r873682170

pikachom committed:
lint: ktLint 수행

pikachom commented:
@wisemuji 
리뷰해주신 코멘트 반영해보았습니다. 감사합니다 (_ _)

wisemuji commented on src/main/kotlin/racing/Car.kt:
```diff
@@ -0,0 +1,10 @@
+package racing
+
+class Car(val number: Int, _currentLocation: Int = 0) {
+    var currentLocation = _currentLocation
+        private set
+
+    fun move() {
+        this.currentLocation++
+    }
+}
```
강의에서 배운 내용을 바로 적용하시는 모습 👍 좋습니다!

wisemuji commented on src/main/kotlin/racing/Application.kt:
```diff
@@ -0,0 +1,10 @@
+package racing
+
+fun main() {
+    val carNumberInput = InputView().showCarInput()
+    val turnNumberInput = InputView().showTurnInput()
```
InputView 인스턴스가 두 번 생성되고 있어요!

pikachom replied:
수정하였습니다!

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,25 @@
+            gameEngine.playTurnForAllCar(carList)
+            ResultView().printCurrentRound(carList)
+        }
+    }
```
아마 나중에 5단계를 진행하시다 보면 다음과 같이 각 패키지별로 의존성을 어떻게 가져가면 좋을지 고민하시게 될텐데요, Game이 ResultView를 꼭 알아야 할지 먼저 고민을 선행해보셔도 좋을 것 같아요!
<img width="993" alt="image" src="https://user-images.githubusercontent.com/32327475/169103352-83df2926-2e8b-4b1c-b6b1-ec31869bacf1.png">

지금 단계에서는 넘어가도 좋습니다!

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,25 @@
+        initializeCarList(carNumber)
+        ResultView().showReadyMessage()
+
+        (0 until turnNumber).forEach { _ ->
```
```suggestion
        repeat(turnNumber) {
```
kotlin [repeat](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/repeat.html) API를 익혀보고 사용해보시면 어떨까요? (이하 동일합니다)

pikachom replied:
반영하였습니다! 유용한 API 소개해주셔서 감사합니다!

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,25 @@
+
+        (0 until turnNumber).forEach { _ ->
+            gameEngine.playTurnForAllCar(carList)
+            ResultView().printCurrentRound(carList)
```
여기서도 ResultView 인스턴스가 두 번 만들어지고 있어요!

pikachom replied:
수정하였습니다.

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,25 @@
+    val carList: MutableList<Car> = mutableListOf()
+    private val gameEngine = GameEngine
+
+    fun run(carNumber: Int, turnNumber: Int) {
```
`~Number` 라는 네이밍을 봤을 때 carNumber가 자동차 개수를 의미하고 turnNumber가 실행할 횟수를 의미하는지 잘 와닿지 않는 것 같아요. 어떤 변수명이 다른 개발자들도 이해하기 쉬운 변수명일까요?

pikachom replied:
코멘트를 보고 고민해보았을때, number가 불필요한 postfix인 것 같아서 제거하고 car, turn 으로 변경해보았습니다.

wisemuji commented on src/main/kotlin/racing/InputView.kt:
```diff
@@ -0,0 +1,26 @@
+    return this.toIntOrNull() ?: throw IllegalArgumentException("올바른 입력값의 형식이 아닙니다")
+}
+
+private fun Int.validateRange(): Int {
+    require(this > 0) { "입력값의 범위는 1 이상입니다" }
+    return this
+}
```
프로그래밍 요구사항에 `핵심 로직을 구현하는 코드와 UI를 담당하는 로직을 구분한다.`가 있는데요,

이건 뷰에서 해야 하는 로직인지 핵심 로직일지 고민해보시면 좋을 것 같아요. 핵심 로직으로 보고 판별을 핵심 로직에서 하도록 변경하면 테스트를 작성할 수 있다는 장점이 있습니다.

pikachom replied:
이 부분은 핵심이라기보다는 view에서만 쓰이는 Util function이라고 생각했었는데요, "Util 클래스를 만드는 대신 클래스 외부에 최상위 함수로 구현하는 방식" 을 적용해보았던 부분입니다. 따로 Util로 분리하는 것이 좋을까요?

pikachom replied:
아래 테스트 코멘트와 함께 고려하여 따로 Util로 분리하였습니다.

wisemuji commented on src/main/kotlin/racing/ResultView.kt:
```diff
@@ -0,0 +1,15 @@
+        println("실행 결과")
+    }
+
+    fun printCurrentRound(carList: MutableList<Car>) {
```
MutableList로 받는건 꼭 필요한 상황이 아니라면 지양하는 편이 좋습니다! 불변 컬렉션을 사용하면 데이터가 중간에 수정되지 않았음을 보장할 수 있습니다.

pikachom replied:
run 시점에 carList를 initialize 하려는 의도에 맞추기 위해 MutableList로 받도록 했었는데요, 말씀해주신 것처럼 불변 컬렉션을 사용하도록 하기 위해 lateinit으로 carList를 멤버변수로 갖도록 하고, initializeCarList에서는 forEach로 add하는 대신 map으로 carList를 변환하는 방식으로 변경해보았습니다.

wisemuji commented on src/test/kotlin/racing/CarTest.kt:
```diff
@@ -0,0 +1,28 @@
+import io.kotest.core.spec.style.FunSpec
+import io.kotest.matchers.shouldBe
+
+class CarTest : FunSpec({
```
저번 단계에서 제안드린 BehaviorSpec을 사용하면 주석 없이도 다음처럼 작성할 수 있습니다!
```kotlin
class CarTest : BehaviorSpec({
    given("blah") {
        val carNumber = 5
        `when`("blah") {
            val carInstance = Car(carNumber)
            then("blah") {
                carInstance.number shouldBe 5
                carInstance.currentLocation shouldBe 0
            }
        }
    }
```

pikachom replied:
제안해주신 시점에 이미 작성중이던 부분이라 FuSpec으로 작성했었습니다 😭 스텝4부터 behaviorspec으로 작성해도 괜찮을까요?

wisemuji replied:
네 그러시죠!!

wisemuji commented on src/test/kotlin/racing/CarTest.kt:
```diff
@@ -0,0 +1,28 @@
+            carInstance.currentLocation shouldBe 0
+        }
+    }
+    context("move 단위테스트") {
```
테스트가 깨집니다!

pikachom replied:
실수로 move를 호출하지 않는 상태에서 커밋했었네요 수정했습니다 😭

wisemuji commented on src/test/kotlin/racing/InputViewTest.kt:
```diff
@@ -0,0 +1,39 @@
+import io.kotest.data.row
+import io.kotest.matchers.shouldBe
+
+class InputViewTest : FunSpec({
```
<img width="550" alt="image" src="https://user-images.githubusercontent.com/32327475/169112083-2d2f5d1f-7243-48a8-8e4a-71072771fcaf.png">
UI 관련 로직은 테스트에서 제외해야 합니다. 핵심 로직과 UI 로직을 분리하면 이 로직도 테스트할 수 있겠죠?

pikachom replied:
InputValidationUtil 이라는 파일을 만들어서 해당 로직을 분리하고, 테스트를 변경하였습니다. 위에 InputView 와 함께 코멘트 모두 반영하였습니다.

wisemuji reviewed (changes requested):
안녕하세요 기훈님!

리뷰를 잘 반영해주셨습니다 💯 고민하신 흔적이 보입니다!
조금 더 고민해보시면 좋을만한 코멘트를 달았으니 확인해주시면 감사하겠습니다.

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

pikachom commented:
코멘트 내용들 반영해보았습니다. 좋은 코멘트 감사합니다.

pikachom committed:
refactor: 2차 코멘트 반영 커밋

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,25 @@
+        val resultView = ResultView()
+        resultView.showReadyMessage()
+
+        repeat(turn){
```
린트가 깨지는 곳들이 있어서 전체적으로 한번 확인해주시면 좋을 것 같습니다!

wisemuji commented on src/main/kotlin/racing/Game.kt:
```diff
@@ -0,0 +1,25 @@
+        carList = listOf()
+    }
+
+    private fun initializeCarList(carNumber: Int) {
+        this.carList = (0 until carNumber)
+            .map{ Car(it) }
+    }
```
carList가 Game 내부에 선언되어 있는데요, Game이 초기화될때 생성자로 필요한 값을 받아 같이 초기화할 수 있지 않을까요? 그렇게 되면 lateinit var를 사용하지 않아도 됩니다. 

참고로 리스트 타입은 var로 선언하여 값을 비울 때 empty list로 초기화하는 것보다 mutable list로 만들어 값을 비우는 것이 더 효율적입니다.

wisemuji commented on src/test/kotlin/racing/CarTest.kt:
```diff
@@ -0,0 +1,29 @@
+import io.kotest.matchers.shouldBe
+
+class CarTest : FunSpec({
+    context("생성자 단위테스트") {
+        test("생성된 car 객체는 입력받은 number와 currentLocation 초기값 0을 가진다") {
+            // given
+            val carNumber = 5
+            // when
+            val carInstance = Car(carNumber)
+            // then
+            carInstance.number shouldBe 5
+            carInstance.currentLocation shouldBe 0
+        }
+    }
```
이건 개발자마다 개인차가 있을 것 같은 의견인데요, 저는 개인적으로 이런 성격의 생성자 테스트는 지양하는 편입니다. 코틀린의 Class를 테스트한다는 생각이 들어서요, 기훈님의 의견은 어떠신가요? 🙂

wisemuji commented on src/test/kotlin/racing/CarTest.kt:
```diff
@@ -0,0 +1,29 @@
+            // given
+            val carInstance = Car(5)
+            // when, then
+            for (i in 0..5) {
+                carInstance.move()
+                carInstance.currentLocation shouldBe i + 1
+            }
```
```suggestion
            for (i in 1..5) {
                carInstance.move()
                carInstance.currentLocation shouldBe i
            }
```
이렇게 작성하면 +1 연산을 없앨 수 있을 것 같습니다 🙂

wisemuji commented on src/test/kotlin/racing/GameEngineTest.kt:
```diff
@@ -0,0 +1,57 @@
+            forAll(
+                row(0, 10, 4),
+                row(1, 5, 8),
+                row(2, 2, 6)
```
forAll , row 사용하여 테스트 진행하신거 넘 좋아요!!!! 👍

wisemuji reviewed (approved):
안녕하세요 기훈님! 리뷰 반영을 잘 해주셨습니다 💯 

조금 고민해보시면 좋을 내용을 코멘트로 달았으니 다음 단계 미션 진행해주시면서 같이 반영해주시면 됩니다!

빠른 미션 진행을 위해 머지하겠습니다. 다음 미션도 화이팅입니다 🙏
